### PR TITLE
Allow for aborting rescans in the GUI

### PR DESCRIPTION
**Pull request by achow101**
A cancel button is added to the `showProgress` dialog that is used only for rescans. When clicked, `AbortRescan` is called directly to cancel the rescan.

Rescans triggered from the debug console will now be cancelable by clicking the cancel button.

Rescans triggered by a command (e.g. `importmulti`) will now give an error indicating that the rescan was aborted by the user (either by the `abortrescan` command or by clicking cancel).

### CODE REVIEW

File: src/qt/bitcoingui.cpp [hunk 1]
```diff
@@ -1119,16 +1120,21 @@ void BitcoinGUI::detectShutdown()
...
     {
         progressDialog = new QProgressDialog(title, "", 0, 100);
         progressDialog->setWindowModality(Qt::ApplicationModal);
         progressDialog->setMinimumDuration(0);
-        progressDialog->setCancelButton(0);
         progressDialog->setAutoClose(false);
         progressDialog->setValue(0);
+
+        if (cancel) {
+            progressDialog->setCancelButtonText("Cancel");
```
Comment (MeshCollider) [inline src/qt/bitcoingui.cpp]: Does this "Cancel" need a translation?

**Review by meshcollider — COMMENTED** [review]
utACK modulo Travis issues

File: src/qt/bitcoingui.cpp [hunk 2]
```diff
@@ -1138,8 +1144,14 @@ void BitcoinGUI::showProgress(const QString &title, int nProgress)
             progressDialog->deleteLater();
         }
     }
-    else if (progressDialog)
-        progressDialog->setValue(nProgress);
+    else if (progressDialog) {
+        if (progressDialog->wasCanceled()) {
+            cancel();
+        }
```
Comment (promag) [inline src/qt/bitcoingui.cpp]: `} else {`

File: src/qt/bitcoingui.cpp [hunk 3]
```diff
@@ -1119,16 +1120,21 @@ void BitcoinGUI::detectShutdown()
     }
 }
 
-void BitcoinGUI::showProgress(const QString &title, int nProgress)
+void BitcoinGUI::showProgress(const QString &title, int nProgress, const std::function<void(void)> &cancel)
```
Comment (promag) [inline src/qt/bitcoingui.cpp]: `&` are next to the type. Same for remaining changes.

**Review by promag — COMMENTED** [review]
Should it be possible to abort rescans caused by the RPC calls?

I'm not very fond of callbacks. Have you thought in alternatives?

**Comment by achow101** [comment]
I don't know what is causing the travis failures.

> Should it be possible to abort rescans caused by the RPC calls?

Yes. There is an `abortrescan` RPC call that lets you abort a rescan. But it is not accessible from the GUI.

File: src/txdb.cpp [hunk 4]
```diff
@@ -386,7 +386,7 @@ bool CCoinsViewDB::Upgrade() {
             if (count++ % 256 == 0) {
                 uint32_t high = 0x100 * *key.second.begin() + *(key.second.begin() + 1);
                 int percentageDone = (int)(high * 100.0 / 65536.0 + 0.5);
-                uiInterface.ShowProgress(_("Upgrading UTXO database") + "\n"+ _("(press q to shutdown and continue later)") + "\n", percentageDone);
+                uiInterface.ShowProgress(_("Upgrading UTXO database") + "\n"+ _("(press q to shutdown and continue later)") + "\n", percentageDone, std::function<void(void)>());
```
Comment (ryanofsky) [inline src/txdb.cpp]: In commit "Add cancel function reference to ShowProgress"

No need to write `std::function<void(void)>()` here and 12 other places. You can just pass {} or nullptr.

File: src/qt/bitcoingui.cpp [hunk 5]
```diff
@@ -1139,8 +1144,14 @@ void BitcoinGUI::showProgress(const QString &title, int nProgress, const std::fu
             progressDialog->deleteLater();
         }
     }
-    else if (progressDialog)
-        progressDialog->setValue(nProgress);
+    else if (progressDialog) {
+        if (progressDialog->wasCanceled()) {
+            cancel();
```
Comment (ryanofsky) [inline src/qt/bitcoingui.cpp]: In commit "Add cancel button to rescan progress dialog"

Maybe assert(cancel)

**Review by ryanofsky — COMMENTED** [review]
utACK e6e8b0c259594b09e405fc9123d1f8473e42b0e4 with RPC bugfixes

File: src/wallet/rpcdump.cpp [hunk 6]
```diff
@@ -535,7 +556,14 @@ UniValue importwallet(const JSONRPCRequest& request)
     file.close();
     pwallet->ShowProgress("", 100, std::function<void(void)>()); // hide progress dialog in GUI
     pwallet->UpdateTimeFirstKey(nTimeBegin);
-    pwallet->RescanFromTime(nTimeBegin, false /* update */);
+    int64_t scanned_time = pwallet->RescanFromTime(nTimeBegin, false /* update */);
+    if (scanned_time > TIMESTAMP_MIN) {
```
Comment (ryanofsky) [inline src/wallet/rpcdump.cpp]: In commit "Give an error when rescan is aborted by the user"

s/TIMESTAMP_MIN/nTimeBegin/ here to fix test failures

File: src/txdb.cpp (hunk 4 above)
Comment (achow101) [reply src/txdb.cpp]: Oh, I didn't realize that that would work. Done.

File: src/qt/bitcoingui.cpp (hunk 3 above)
Comment (achow101) [reply src/qt/bitcoingui.cpp]: Done

File: src/qt/bitcoingui.cpp (hunk 2 above)
Comment (achow101) [reply src/qt/bitcoingui.cpp]: Done

File: src/wallet/rpcdump.cpp (hunk 6 above)
Comment (achow101) [reply src/wallet/rpcdump.cpp]: Done

File: src/qt/bitcoingui.cpp (hunk 1 above)
Comment (achow101) [reply src/qt/bitcoingui.cpp]: Yes, it does. Done

**Review by ryanofsky — COMMENTED** [review]
utACK c8720f0a320eff0cc83a459dc87f28087d518825. Only changes were implementing various suggestions.

> I'm not very fond of callbacks. Have you thought in alternatives?

Reason callback seems to be used is to avoid the need for qt code to call CWallet's AbortRescan method or set the fAbortRescan flag directly. The alternatives would be for Qt to call a different function or set a different flag, but don't seem like they would be much better or worse.

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -1568,13 +1568,13 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool f
         fAbortRescan = false;
         fScanningWallet = true;
 
-        ShowProgress(_("Rescanning..."), 0); // show rescan progress in GUI as dialog or on splashscreen, if -rescan on startup
+        ShowProgress(_("Rescanning..."), 0, boost::bind(&CWallet::AbortRescan, this)); // show rescan progress in GUI as dialog or on splashscreen, if -rescan on startup
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: nit: use `std::bind` (and below)

File: src/ui_interface.h [hunk 8]
```diff
@@ -95,7 +95,7 @@ class CClientUIInterface
     boost::signals2::signal<void (CWallet* wallet)> LoadWallet;
 
     /** Show progress e.g. for verifychain */
-    boost::signals2::signal<void (const std::string &title, int nProgress)> ShowProgress;
+    boost::signals2::signal<void (const std::string &title, int nProgress, const std::function<void(void)>& cancel)> ShowProgress;
```
Comment (jonasschnelli) [inline src/ui_interface.h]: use `nullptr` as default parameter for `cancel` to avoid adding `nullptr` to all non-callback `ShowProgress` calls?

**Review by jonasschnelli — COMMENTED** [review]
Concept ACK

File: src/ui_interface.h (hunk 8 above)
Comment (achow101) [reply src/ui_interface.h]: I couldn't set a default parameter in the signal, so I'm not sure it is possible to do that. Or I'm just doing something wrong.

File: src/wallet/wallet.cpp (hunk 7 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Done

**Comment by jonasschnelli** [comment]
Tested a bit and there is a concurrency issue (maybe not related to this code but since you finally can test multiple rescans during the same runtime this PR may reveal the issue):

Sometimes the progress window does not show up (happened to me only if I canceled a rescan before) and the GUI is fully in locked state (due to the rescan).
In that case, the rescan is happening and the GUI is locked (without showing a progress screen) via `cs_main` in `transactiontablemodel.cpp`

**Comment by jonasschnelli** [comment]
Needs rebase

File: src/qt/bitcoingui.cpp [hunk 9]
```diff
@@ -487,7 +487,7 @@ void BitcoinGUI::setClientModel(ClientModel *_clientModel)
         connect(_clientModel, SIGNAL(message(QString,QString,unsigned int)), this, SLOT(message(QString,QString,unsigned int)));
 
         // Show progress dialog
-        connect(_clientModel, SIGNAL(showProgress(QString,int)), this, SLOT(showProgress(QString,int)));
+        connect(_clientModel, SIGNAL(showProgress(QString,int, bool, std::function<void(void)>)), this, SLOT(showProgress(QString,int, std::function<void(void)>)));
```
Comment (ryanofsky) [inline src/qt/bitcoingui.cpp]: Could just write `void()` instead of `void(void)` everywhere

File: src/qt/splashscreen.cpp [hunk 10]
```diff
@@ -181,7 +181,7 @@ static void ShowProgress(SplashScreen *splash, const std::string &title, int nPr
 #ifdef ENABLE_WALLET
 void SplashScreen::ConnectWallet(CWallet* wallet)
 {
-    wallet->ShowProgress.connect(boost::bind(ShowProgress, this, _1, _2, false));
+    wallet->ShowProgress.connect(boost::bind(ShowProgress, this, _1, _2, false, _4));
```
Comment (ryanofsky) [inline src/qt/splashscreen.cpp]: You added both `resume_possible` and `cancel` arguments to the ShowProgress signal in wallet.h so this should be `_3, _4`, instead of `false, _4`.

(Another alternative if you want to ensure wallet resume_possible value is always false, is to not add resume_possible in wallet.h, and then change this to `false, _3`. But adding the argument and then silently ignoring like this does now isn't good.)

**Review by ryanofsky — COMMENTED** [review]
utACK 4657e40f1516d506095f969dc726da6f97d57a99. Only change since last review is rebase and some  whitespace diffs.

**Comment by promag** [comment]
utACK 4657e40. Have you thought adding tests for the new RPC errors? (might be impossible?)

**Comment by jonasschnelli** [comment]
Reminder: this PR currently causes a deadlock (at least on OSX):
IMO it would require https://github.com/bitcoin/bitcoin/pull/11281 to work flawless.

**Comment by ryanofsky** [comment]
@jonasschnelli, I don't see how this could be causing any deadlock that wasn't occurring before. The cancel callback isn't acquiring any locks, just updating an atomic\<bool>. Are you sure there is a problem in the current version of this PR?

**Comment by achow101** [comment]
@ryanofsky I think it is more that a deadlock currently exists in the rescanning code which this PR makes more obvious and easier to hit.

File: src/qt/bitcoingui.cpp (hunk 9 above)
Comment (achow101) [reply src/qt/bitcoingui.cpp]: Done

File: src/qt/splashscreen.cpp (hunk 10 above)
Comment (achow101) [reply src/qt/splashscreen.cpp]: Done

Comment (achow101) [reply src/qt/splashscreen.cpp]: Actually changing this seems to be causing a segfault, so I will leave it as is.

**Comment by ryanofsky** [comment]
> @ryanofsky I think it is more that a deadlock currently exists in the rescanning code which this PR makes more obvious and easier to hit.

I know that's the claim, but I don't understand what the deadlock is. What does the rescan code have to do with transactiontablemodel and why would shorter aborted scans cause more UI problems than longer full scans?

**Review by ryanofsky — COMMENTED** [review]
utACK 449dd52b417d2a49b4de4440b391f6ebccb093b4. Same as before, just gets rid of spurious whitespace changes and void(void)

**Comment by promag** [comment]
Please see an alternative concept in #11826.

**Comment by achow101** [comment]
Rebased

**Comment by achow101** [comment]
Review/Merge beg

**Comment by promag** [comment]
@achow101 needs rebase.

**Comment by achow101** [comment]
Rebased

File: src/qt/bitcoingui.h [hunk 11]
```diff
@@ -232,8 +234,8 @@ private Q_SLOTS:
     void detectShutdown();
 
     /** Show progress dialog e.g. for verifychain */
-    void showProgress(const QString &title, int nProgress);
-    
+    void showProgress(const QString &title, int nProgress, bool resume_possible, const std::function<void()>& cancel);
```
Comment (luke-jr) [inline src/qt/bitcoingui.h]: Would be nice to have nullptr as a default for cancel.

File: src/wallet/rpcdump.cpp [hunk 12]
```diff
@@ -608,10 +628,18 @@ UniValue importwallet(const JSONRPCRequest& request)
...
         file.close();
-        pwallet->ShowProgress("", 100); // hide progress dialog in GUI
+        pwallet->ShowProgress("", 100, false, nullptr); // hide progress dialog in GUI
         pwallet->UpdateTimeFirstKey(nTimeBegin);
     }
-    pwallet->RescanFromTime(nTimeBegin, reserver, false /* update */);
+    pwallet->ShowProgress("", 100, false, nullptr); // hide progress dialog in GUI
+    int64_t scanned_time = pwallet->RescanFromTime(nTimeBegin, reserver, false /* update */);
+    if (scanned_time > nTimeBegin) {
+        if (pwallet->IsAbortingRescan()) {
```
Comment (luke-jr) [inline src/wallet/rpcdump.cpp]: This is ugly, since RescanFromTime changes without a signature change. Better to have it throw an exception to be caught from callers, when cancelled.

Comment (achow101) [reply src/wallet/rpcdump.cpp]: I don't quite understand what you mean.

File: src/qt/bitcoingui.h (hunk 11 above)
Comment (achow101) [reply src/qt/bitcoingui.h]: This was suggested before and I was not able to get it to work.

**Comment by ryanofsky** [comment]
utACK 7e53d957709c0bac88cabfc0f9370b33888504db. No changes since last review in Qt code. Just conflict resolution in other parts of the code.

**Comment by achow101** [comment]
Review beg for reviews from people other than @ryanofsky

File: src/wallet/rpcdump.cpp (hunk 12 above)
Comment (sipa) [reply src/wallet/rpcdump.cpp]: @luke-jr In what way does `RescanFromTime` change?

**Comment by MarcoFalke** [comment]
utACK, but calling `rescanblockchain` in the GUI console no longer gives me the tiny "Rescanning ..." pop up, which prevents me from testing. Any idea?

Pop up on master:

screenshot from 2018-03-18 09-10-58

**Comment by achow101** [comment]
That's strange. I'm seeing that now too, but I don't know why.

**Comment by achow101** [comment]
Ok, I fixed that problem and also had to rebase.

File: src/qt/splashscreen.cpp (hunk 10 above)
Comment (MarcoFalke) [reply src/qt/splashscreen.cpp]: The following diff appears to be working for me. Could you help me find the steps to reproduce the segfault? Otherwise, I think this is a sensible suggestion.

```diff
--- a/src/qt/splashscreen.cpp
+++ b/src/qt/splashscreen.cpp
@@ -181,7 +181,7 @@ static void ShowProgress(SplashScreen *splash, const std::string &title, int nPr
 #ifdef ENABLE_WALLET
 void SplashScreen::ConnectWallet(CWallet* wallet)
 {
-    wallet->ShowProgress.connect(boost::bind(ShowProgress, this, _1, _2, false, _4));
+    wallet->ShowProgress.connect(boost::bind(ShowProgress, this, _1, _2, _3, _4));
     connectedWallets.push_back(wallet);
 }
 #endif
@@ -203,7 +203,7 @@ void SplashScreen::unsubscribeFromCoreSignals()
     uiInterface.ShowProgress.disconnect(boost::bind(ShowProgress, this, _1, _2, _3, _4));
 #ifdef ENABLE_WALLET
     for (CWallet* const & pwallet : connectedWallets) {
-        pwallet->ShowProgress.disconnect(boost::bind(ShowProgress, this, _1, _2, false, _4));
+        pwallet->ShowProgress.disconnect(boost::bind(ShowProgress, this, _1, _2, _3, _4));
     }
 #endif
 }

File: src/wallet/rpcdump.cpp (hunk 12 above)
Comment (MarcoFalke) [reply src/wallet/rpcdump.cpp]: `RescanFromTime` uses `ScanForWalletTransactions` internally. And the interface of `ScanForWalletTransactions` is changed in this pull request.

File: src/wallet/wallet.cpp [hunk 13]
```diff
@@ -1773,6 +1773,7 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
         }
         if (pindex && fAbortRescan) {
             LogPrintf("Rescan aborted at block %d. Progress=%f\n", pindex->nHeight, gvp);
+            ret = pindex;
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: Please revert this change. This breaks other rpcs that use `ScanForWalletTransactions` directly

File: src/wallet/rpcdump.cpp [hunk 14]
```diff
@@ -479,7 +493,14 @@ UniValue importpubkey(const JSONRPCRequest& request)
     }
     if (fRescan)
     {
-        pwallet->RescanFromTime(TIMESTAMP_MIN, reserver, true /* update */);
+        int64_t scanned_time = pwallet->RescanFromTime(TIMESTAMP_MIN, reserver, true /* update */);
+        if (scanned_time > TIMESTAMP_MIN) {
+            if (pwallet->IsAbortingRescan()) {
```
Comment (MarcoFalke) [inline src/wallet/rpcdump.cpp]: Why is this check within the scope of the `(scanned_time > TIMESTAMP_MIN)` check? Wouldn't the following make more sense:

```cpp
if (pwallet->IsAbortingRescan()) throw JSONRPCError(RPC_MISC_ERROR, "Rescan aborted by user.");
if (scanned_time > TIMESTAMP_MIN) throw JSONRPCError(RPC_WALLET_ERROR, "Rescan was unable to fully rescan the blockchain. Some transactions may be missing.");

**Comment by MarcoFalke** [comment]
utACK c18cfb8fe0e5de02471011bef0e37f0a6d5920e6, but needs feedback addressed.

File: src/wallet/wallet.cpp (hunk 13 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Done

File: src/wallet/rpcdump.cpp (hunk 14 above)
Comment (achow101) [reply src/wallet/rpcdump.cpp]: Yes, done.

File: src/wallet/rpcdump.cpp (hunk 12 above)
Comment (achow101) [reply src/wallet/rpcdump.cpp]: That change was reverted.

File: src/qt/splashscreen.cpp (hunk 10 above)
Comment (achow101) [reply src/qt/splashscreen.cpp]: Hmm, that diff works for me. I don't remember what I did that caused a segfault. Added this change.

**Comment by MarcoFalke** [comment]
re-utACK 1de7f88f4cad0f59a6684e179177707c8efd7b3a (only changes were to address the feedback)

**Review by ryanofsky — COMMENTED** [review]
utACK 1de7f88f4cad0f59a6684e179177707c8efd7b3a. Changes since last review were rebase, checking for abort condition even if RescanFromTime succeeds, and forwarding `resume_possible` argument to `showProgress`.

**Comment by MarcoFalke** [comment]
@jonasschnelli You tested this a bit last year. Mind to re-test?

**Comment by TheBlueMatt** [comment]
This effectively reverts #10770. Please, no.

**Comment by achow101** [comment]
@TheBlueMatt I don't see any other way to implement such cancellations in the GUI. Perhaps you could suggest something?

**Comment by TheBlueMatt** [comment]
We cancel just fine during init already (using StartShutdown()). If we want to *also* support cancelling via a rescan progress dialog, the rescan progress dialog should know how to make that happen. I still really don't think we need a generic interface here - we only use the dialog for one thing (and I'm not aware of any other cancelable proposed things to show there) - and making it common between init and rescan still makes no sense to me - they are very, very different things.

**Comment by ryanofsky** [comment]
@TheBlueMatt , your objection seems the same as the "I'm not very fond of callbacks" criticism which came up earlier: https://github.com/bitcoin/bitcoin/pull/11200#pullrequestreview-59957173. I didn't think this was a big deal (or really a very substantive criticism), but it would be pretty easy to address by replacing the callback with an wallet `AbortRescans()` function that Qt code could call directly. Would this make everybody happy?

**Comment by TheBlueMatt** [comment]
Its a bit more broad, I'm a bit more than "not fond of" passing callbacks from random internal places all the way to GUI, plus we started with that and deliberately got rid of them. Adding them *back* would need more justification than "well, this is a way that it can be done".

**Comment by ryanofsky** [comment]
I think support cancelling rescans is a worthwhile feature, and there are lots of ways it can be implemented. Which way to choose is mostly a question of how much much existing code you want to rewrite.

Having qt code just call an `AbortRescan` function when the cancel button is pressed is probably the simplest possible approach. Having the qt code call a generic `cancel` callback is a little more generic, and maybe a little cleaner, but does involve callbacks. Changing the progress dialog to be aware of the differences between init and rescan would be another way to approach this, and probably a good idea, but just a larger change.

**Comment by achow101** [comment]
I changed the implementation to completely remove the callback. The `showProgress` dialog that is used for rescans just calls `AbortRescan` directly when cancel is clicked. This should work since the dialog is used only for rescanning.

**Comment by achow101** [comment]
Rebased

**Comment by MarcoFalke** [comment]
Needs trivial rebase to accommodate a file rename.

**Comment by achow101** [comment]
Rebased

File: src/wallet/rpcdump.cpp [hunk 15]
```diff
@@ -172,7 +172,13 @@ UniValue importprivkey(const JSONRPCRequest& request)
         }
     }
     if (fRescan) {
-        pwallet->RescanFromTime(TIMESTAMP_MIN, reserver, true /* update */);
+        int64_t scanned_time = pwallet->RescanFromTime(TIMESTAMP_MIN, reserver, true /* update */);
```
Comment (promag) [inline src/wallet/rpcdump.cpp]: This code is repeated 4 times, maybe it's worth moving it to a function?

File: src/interfaces/wallet.h [hunk 16]
```diff
@@ -247,6 +247,9 @@ class Wallet
     //! Register handler for watchonly changed messages.
     using WatchOnlyChangedFn = std::function<void(bool have_watch_only)>;
     virtual std::unique_ptr<Handler> handleWatchOnlyChanged(WatchOnlyChangedFn fn) = 0;
+
+    //! Abort a rescan
```
Comment (promag) [inline src/interfaces/wallet.h]: Nit, period.

File: src/qt/walletmodel.cpp [hunk 17]
```diff
@@ -558,3 +558,8 @@ bool WalletModel::isMultiwallet()
 {
     return m_node.getWallets().size() > 1;
 }
+
+void WalletModel::AbortRescan() const
+{
+    wallet().AbortRescan();
```
Comment (promag) [inline src/qt/walletmodel.cpp]: Nit, `m_wallet->AbortRescan()`?

File: src/qt/walletmodel.h [hunk 18]
```diff
@@ -204,6 +204,8 @@ class WalletModel : public QObject
     QString getWalletName() const;
 
     bool isMultiwallet();
+
+    void AbortRescan() const;
```
Comment (promag) [inline src/qt/walletmodel.h]: Nit, `void abortRescan() const`.

**Review by promag — COMMENTED** [review]
Some comments, will test later.

**Comment by promag** [comment]
> Changing the progress dialog to be aware of the differences between init and rescan would be another way to approach this, and probably a good idea, but just a larger change.

I agree with @ryanofsky here. I'll try to push #11826 forward.

File: src/interfaces/wallet.h (hunk 16 above)
Comment (ryanofsky) [reply src/interfaces/wallet.h]: Would suggest renaming this method `abortRescan` and moving it above `backupWallet` (both here and in `wallet.cpp`). These interfaces are organized with regular methods on top and callback methods below. Also roughly with whole-wallet operations above operations on individual addresses and transactions.

**Comment by jonasschnelli** [comment]
Could finally test this.
I think it is an important feature because the way how users currently abort rescan is by force quitting the application (which is really not what we want).

Tested ACK bc2003cdf7102e03710dbc19e9c397b92e124fb4.
I'd like to see @promag's nits addresses (lowercase methodname in QT, interface call via class member).

File: src/qt/walletmodel.cpp (hunk 17 above)
Comment (ryanofsky) [reply src/qt/walletmodel.cpp]: In commit "Add cancel button to rescan progress dialog"

Would suggest dropping this method and just calling model.wallet().abortRescan() directly. I don't think there's a benefit to wrapping individual `Wallet` calls inside a `WalletModel` methods unless they're doing extra work like synchronization or caching.

File: src/wallet/rpcdump.cpp (hunk 15 above)
Comment (ryanofsky) [reply src/wallet/rpcdump.cpp]: In commit "Give an error when rescan is aborted by the user"

> This code is repeated 4 times, maybe it's worth moving it to a function?

Could be nice, though to be honest a lot of other code is duplicated between the import* RPCs too. So effort might be better spent deduping larger chunks or deprecating older RPCs in favor of importmulti.

**Review by ryanofsky — COMMENTED** [review]
utACK bc2003cdf7102e03710dbc19e9c397b92e124fb4. This PR has changed quite a lot since it was opened but is very simple at this point and provides useful functionality.

File: src/wallet/rpcdump.cpp (hunk 15 above)
Comment (achow101) [reply src/wallet/rpcdump.cpp]: I think that should be part of a separate refactor for deduplicating all of the import* RPCs.

File: src/interfaces/wallet.h (hunk 16 above)
Comment (achow101) [reply src/interfaces/wallet.h]: Done (moved and added period).

File: src/qt/walletmodel.cpp (hunk 17 above)
Comment (achow101) [reply src/qt/walletmodel.cpp]: Removed and called directly from walletview

File: src/qt/walletmodel.h (hunk 18 above)
Comment (achow101) [reply src/qt/walletmodel.h]: Removed

File: src/wallet/rpcdump.cpp [hunk 19]
```diff
@@ -534,9 +534,9 @@ UniValue importwallet(const JSONRPCRequest& request)
         int64_t nFilesize = std::max((int64_t)1, (int64_t)file.tellg());
         file.seekg(0, file.beg);
 
-        pwallet->ShowProgress(_("Importing..."), 0); // show progress dialog in GUI
+        uiInterface.ShowProgress(_("Importing..."), 0, false); // show progress dialog in GUI
```
Comment (ryanofsky) [inline src/wallet/rpcdump.cpp]: In commit "Add cancel button to rescan progress dialog"

These changes from wallet to global `ShowProgress` calls seem unrelated. Are they intentional? It might make sense to change these if removing the wallet `ShowProgress` signal, but since that is still used in `CWallet::ScanForWalletTransactions`, this change just makes these calls inconsistent.

If this change is necessary it would be good to have a comment saying why CWallet::ShowProgress is not used here.

**Review by ryanofsky — COMMENTED** [review]
utACK 66ee39f227b28a32c6c1cb6cd68c0eee0cc8637f

File: src/wallet/rpcdump.cpp (hunk 19 above)
Comment (achow101) [reply src/wallet/rpcdump.cpp]: Because `CWallet::ShowProgress` has a cancel button now which is directly tied to `AbortRescan` (because it isn't used for anything except showing rescan progress), I needed a different progress bar which does not have a cancel button. This change makes that `ShowProgress` use `BitcoinGUI::ShowProgress` (which is tied toe `uiInterface`)which does not have a cancel button (and is also not used anywhere else AFAICT). This progress bar indicates the import progress, not rescan progress which is why a different `ShowProgress` is needed

**Review by ryanofsky — COMMENTED** [review]
utACK deaffd812e3cdaeb7ec978c2d86f4bf688add96f, thanks for the comment!

**Comment by achow101** [comment]
Squashed the squashme. Is this ready for merging?

**Comment by MarcoFalke** [comment]
utACK ae1d2b0308d7fe9df7fc18699c69e9587e6154bd. Going to test this later...

File: src/wallet/rpcdump.cpp (hunk 15 above)
Comment (promag) [reply src/wallet/rpcdump.cpp]: Sure.
